Support condition always true in prefix unary expression

tjenkinson opened the pull request:
<!--
Thank you for submitting a pull request!

Please verify that:
* [ ] There is an associated issue in the `Backlog` milestone (**required**)
* [ ] Code is up-to-date with the `main` branch
* [ ] You've successfully run `hereby runtests` locally
* [ ] There are new or updated unit tests validating the change

Refer to CONTRIBUTING.MD for more details.
  https://github.com/Microsoft/TypeScript/blob/main/CONTRIBUTING.md

** Please don't send typo fixes! **
Please don't send a PR solely for the purpose of fixing a typo, unless that
typo truly hurts understanding of the text. Each PR represents work for the
maintainers, and that work should provide commensurate value.

If you're interested in sending a PR, the issue tracker has many issues marked `help wanted`.
-->

Opening as draft for now so we can run the extended tests and see how much breakage there is. See https://github.com/microsoft/TypeScript/issues/46140#issuecomment-2375375422

cc @gabritto

Fixes #46140

gabritto commented:
@typescript-bot test it

gabritto commented:
As I suspected, there are a lot of new errors in our extended tests.

I looked at a few of the new errors for *uncalled functions*, and there it seems that indeed, to support erroring on negated function conditions, we'd have to do the analysis of whether the function is called  outside the `if` block, similar to what we do currently for non-negated function conditions to avoid false errors. I still have to look at the rest of the new function errors.

For the new *promise* errors, apollographql/apollo-client error looks like an error that is caused by an incorrect non-null assertion, so the new error is fine and could be fixed by them getting rid of the non-null assertion. For the blakeblackshear/frigate error, it looks like a bad new error because the promise is awaited and possibly we'd have to do a similar analysis of "is this promise awaited afterwards?". In vscode, https://github.com/microsoft/vscode/blob/f71675cbd9ba6aef444321cbe5faf8dd81340e3b/src/vs/base/node/id.ts#L82, https://github.com/microsoft/vscode/blob/f71675cbd9ba6aef444321cbe5faf8dd81340e3b/src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts#L526, https://github.com/microsoft/vscode/blob/f71675cbd9ba6aef444321cbe5faf8dd81340e3b/src/vs/workbench/services/search/common/search.ts#L819, https://github.com/microsoft/vscode/blob/f71675cbd9ba6aef444321cbe5faf8dd81340e3b/extensions/html-language-features/server/src/htmlServer.ts#L121, https://github.com/microsoft/vscode/blob/f71675cbd9ba6aef444321cbe5faf8dd81340e3b/extensions/emmet/src/test/abbreviationAction.test.ts#L223 are false errors, https://github.com/microsoft/vscode/blob/f71675cbd9ba6aef444321cbe5faf8dd81340e3b/src/vs/workbench/browser/layout.ts#L1759 is maybe a good error, https://github.com/microsoft/vscode/blob/f71675cbd9ba6aef444321cbe5faf8dd81340e3b/extensions/gulp/src/main.ts#L184 is a good error.
https://github.com/video-dev/hls.js/blob/5cfbc0a0ce5587fa2a49b607d1e95861b20ac5cb/tests/unit/controller/eme-controller.ts#L228 I don't even know how to interpret, and https://github.com/video-dev/hls.js/blob/5cfbc0a0ce5587fa2a49b607d1e95861b20ac5cb/tests/unit/controller/eme-controller.ts#L447 is a false error that I'm not even sure how we could avoid...

For desktop/desktop, gitpod-io/gitpod, the new errors are good and I think they all indicate bugs.

gabritto commented:
@tjenkinson I think both the function and promise cases will require some extra work for avoiding most of the false errors. Let me know if you'd still like to keep working on it, I think it's a very good feature since we had at least 3 bugs detected in the extended tests.

tjenkinson commented:
hey @gabritto thanks for looking through them super promising it found some actual issues.

Do you think we'll need to handle cases like [this one](https://github.com/microsoft/vscode/blob/f71675cbd9ba6aef444321cbe5faf8dd81340e3b/extensions/emmet/src/test/abbreviationAction.test.ts#L223) where the usage happens later and is not part of the `else` block? Or would just checking the else block be good enough.

Happy to continue digging but I think that will be a bit more complex looking at how things are structured at the moment as `block` can't just be the `else` body. It would need to traverse up to the containing block, find the `if/else`, and then look in all items below it?

tjenkinson commented:
Actually maybe this is a different kind of check.

If there's an `if (!a)`, it needs to be ignored if:
1. inside the `if` block something is assigned to `a`
2. there's an `else` that uses `a`

I'm not sure if (2) will happen much in the wild so maybe the 1st will be enough

tjenkinson commented:
Also not sure https://github.com/microsoft/vscode/blob/f71675cbd9ba6aef444321cbe5faf8dd81340e3b/extensions/emmet/src/test/abbreviationAction.test.ts#L223 is a false error? Looking at the types `expandPromise` should always be defined so that check looks redundant?

gabritto commented:
> Also not sure https://github.com/microsoft/vscode/blob/f71675cbd9ba6aef444321cbe5faf8dd81340e3b/extensions/emmet/src/test/abbreviationAction.test.ts#L223 is a false error? Looking at the types `expandPromise` should always be defined so that check looks redundant?

I think you are right in this case. However, I think we should still support that case to avoid false errors in code written in a similar way.

tjenkinson commented on tests/baselines/reference/contextualOverloadListFromArrayUnion.errors.txt:
```diff
@@ -0,0 +1,69 @@
+    
+        function next() {
+            let task = tasks[index]
+            if (!task) {
```
Looks like might also have to bail if the source is accessing something from an array 🤔

gabritto replied:
This case looks like a bug in your PR code: we should not be erroring because `task` is used inside the `else` block.

tjenkinson replied:
Currently we check for assignment to `task` though. If we just check for it appearing then I think this would be fine

tjenkinson commented:
Hey @gabritto I've pushed some changes so that now it _should_ only apply if there's a single `!` (given that `!!` appears to be a way to opt out of the behaviour for the positive case), and nothing is assigned to the variable inside if the `if` block.

Still need to add tests, but would be great if you could run the run the extended tests again and we could see if this reduces the false positives?

Also might need to add some logic to bail if it's accessing an item from an array

gabritto commented:
@typescript-bot test it

tjenkinson commented:
1. https://github.com/webpack/webpack/blob/3919c844eca394d73ca930e4fc5506fb86e2b094/lib/NormalModule.js#L831 think think this is fine, `identifier` _should_ be optional
2. https://github.com/webpack/webpack/blob/3919c844eca394d73ca930e4fc5506fb86e2b094/lib/dependencies/LoaderPlugin.js#L277 it's a js file and doesn't infer that `callback` could be undefined. I think this is fine? They add a jsdoc type or another ts-ignore 
3. https://github.com/DefinitelyTyped/DefinitelyTyped/blob/de59f128691d22d9bf25ed2b659acae6c50629b1/types/morgan/morgan-tests.ts#L88 is interesting. It's failing because even though it's `let fn: morgan.FormatFn | undefined = developmentExtendedFormatLine.memoizer[color];` it's thinking it can't be `undefined` given it assumes the array access will always find something. I think `let fn = developmentExtendedFormatLine.memoizer[color] as morgan.FormatFn | undefined;` would fix this
4. https://github.com/DefinitelyTyped/DefinitelyTyped/blob/de59f128691d22d9bf25ed2b659acae6c50629b1/types/activex-excel/activex-excel-tests.ts#L98. Not sure. Maybe the types or test is wrong
5. https://github.com/DefinitelyTyped/DefinitelyTyped/blob/de59f128691d22d9bf25ed2b659acae6c50629b1/types/xdate/xdate-tests.ts#L1177 is expected. If the check was inverted it would be fine 🤔 
6. https://github.com/DefinitelyTyped/DefinitelyTyped/blob/de59f128691d22d9bf25ed2b659acae6c50629b1/types/js-fixtures/js-fixtures-tests.ts#L16 maybe the types are wrong, or this is a case where would opt out with `!!!` or use `as`
7. https://github.com/DefinitelyTyped/DefinitelyTyped/blob/de59f128691d22d9bf25ed2b659acae6c50629b1/types/yandex-maps/yandex-maps-tests.ts#L162 is expected. If the check was inverted it would be fine
8. https://github.com/DefinitelyTyped/DefinitelyTyped/blob/de59f128691d22d9bf25ed2b659acae6c50629b1/types/activex-shell/activex-shell-tests.ts#L426. This one shouldn't be failing. Need to figure out why
9. https://github.com/adobe/react-spectrum/blob/4dd0993da6421c14a632b7d0a11b17e759eef38c/packages/dev/codemods/src/index.ts#L86 is expected. If the check was inverted it would be fine
10. https://github.com/baidu/amis/blob/2747feea574344346afdc53940c42547e98f91c7/packages/amis-formula/src/evalutor.ts#L74 is expected. If the check was inverted it would be fine
11. https://github.com/baidu/amis/blob/2747feea574344346afdc53940c42547e98f91c7/packages/amis-formula/src/evalutor.ts#L118  is expected. If the check was inverted it would be fine
12. https://github.com/blakeblackshear/frigate/blob/a1ce9aacf27591d141619e887579c963df55ce8e/web/src/components/player/WebRTCPlayer.tsx#L109 looks like a real bug. I think the type should be `Promise<RTCPeerConnection> | undefined` instead of `Promise<RTCPeerConnection | undefined>`
13. https://github.com/cheeriojs/cheerio/blob/418e98b0f448a7ff470055a3d0cbe61e72605d41/src/static.ts#L26  is expected. If the check was inverted it would be fine
14. https://github.com/clash-verge-rev/clash-verge-rev/blob/ae5b2cfb79423c7e76a281725209b812774367fa/src/components/setting/mods/password-input.tsx#L23 looks like a mistake
15. https://github.com/compiler-explorer/compiler-explorer/blob/b97cbb33b9605082c326cfe7d22aa577193754e8/lib/exec.ts#L412 is expected. If the check was inverted it would be fine
16. https://github.com/compiler-explorer/compiler-explorer/blob/b97cbb33b9605082c326cfe7d22aa577193754e8/lib/exec.ts#L630 is expected. If the check was inverted it would be fine
17. https://github.com/compiler-explorer/compiler-explorer/blob/b97cbb33b9605082c326cfe7d22aa577193754e8/lib/exec.ts#L412 is expected. If the check was inverted it would be fine
18. https://github.com/compiler-explorer/compiler-explorer/blob/b97cbb33b9605082c326cfe7d22aa577193754e8/lib/exec.ts#L630 is expected. If the check was inverted it would be fine

.. started skimming through the rest

20. https://github.com/desktop/desktop/blob/80ac47dc75d308d562c45d97c31dfee86226d3f4/app/src/lib/git/cherry-pick.ts#L219 looks like real bug
21. https://github.com/desktop/desktop/blob/80ac47dc75d308d562c45d97c31dfee86226d3f4/app/src/lib/stores/app-store.ts#L7248 looks like real bug
22. https://github.com/ether/etherpad-lite/blob/13bc06febea15b6d816a125d4c2ca7ceedcf4c58/src/static/js/caretPosition.ts#L194 is expected. If the check was inverted it would be fine
23. https://github.com/gitpod-io/gitpod/blob/4c25f426e1d1c911099e51ad3771143ca380f32c/components/gitpod-db/src/typeorm/migration/helper/helper.ts#L19 looks like a real bug

(5) (7) (9) (10) (11) (13) (15) (16) (17) (18) (22) would be fine if we check if the variable is called/awaited/assigned to anywhere inside the enclosing function

(3) (10) (15) (16) (17) (18) are issues where array access is not also `undefined`

So I think we'll need to do something like

> would be fine if we check if the variable is called/awaited/assigned to anywhere inside the enclosing function

?

Wondering if you have any other ideas or if there could be a performance issue with this? @gabritto

gabritto commented on src/compiler/checker.ts:
```diff
@@ -44588,8 +44593,10 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
+                || isUsedInBody;
+
             if (!isUsed) {
                 if (isPromise) {
```
Why do we call `isSymbolUsedInConditionBody` with `location` instead of `condExpr`?

gabritto commented on src/compiler/checker.ts:
```diff
@@ -44606,15 +44613,23 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     }
 
-    function isSymbolUsedInConditionBody(expr: Expression, body: Statement | Expression, testedNode: Node, testedSymbol: Symbol): boolean {
+    function isSymbolUsedInConditionBody(expr: Expression, body: Statement | Expression | Node, testedNode: Node, testedSymbol: Symbol, checkForAssignment: boolean): boolean {
```
What's the idea here behind `checkForAssignment`? Why were the previous checks being done here in `isSymbolUsedInConditionBody` not enough?

tjenkinson replied:
I think they would probably be fine but then

```ts
if (!a) {
  console.log(a)
}
```

wouldn't error when it probably should? But also it would be weird to reference something after checking it's falsey so maybe it would be fine and simpler just to reuse

gabritto commented on src/compiler/checker.ts:
```diff
@@ -44538,21 +44538,26 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
-            const location = isLogicalOrCoalescingBinaryExpression(condExpr) ? skipParentheses(condExpr.right) : condExpr;
+            let location = condExpr;
+            let inverted = false;
+            if (isPrefixUnaryExpression(location)) {
```
I think you also need to check that the operator is the exclamation token. Otherwise `location` could be another kind of expression, e.g. `++a`.

gabritto commented on src/compiler/checker.ts:
```diff
@@ -44538,21 +44538,26 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
+            let inverted = false;
+            if (isPrefixUnaryExpression(location)) {
+                location = skipParentheses(location.operand);
+                inverted = true;
```
Another thing to consider (and probably add a test) is that we can have an expression that flips between the negative and positive case, if we have nested negation expressions, e.g.: `if (a && !(b || !c)) ...`

tjenkinson replied:
ah 🙈

tjenkinson replied:
I don't think there's really any special handling needed here?

gabritto reviewed (commented):
I left some questions, but also, can you add some tests so we can track which cases are currently supported and which are not, and which are potential false errors? Otherwise, it's hard for me to remember in between reviews.

tjenkinson commented on src/compiler/checker.ts:
```diff
@@ -44588,8 +44593,10 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 return;
             }
 
+            const isUsedInBody = inverted && isIfStatement(condExpr.parent) ? testedSymbol && isSymbolUsedInConditionBody(location, condExpr.parent.parent, testedNode, testedSymbol, /*checkForAssignment*/ true) : testedSymbol && body && isSymbolUsedInConditionBody(condExpr, body, testedNode, testedSymbol, /*checkForAssignment*/ false);
```
For some reason without this it wouldn't detect `if (!something)`, but still detect `if (!something.else)`. Didn't dig much into why yet

tjenkinson replied:
I think it probably should be `location` and before was a potential bug?

For the `inverted` case and `if (!someVar)` we need to pass in the resolved `someVar` not `!someVar` so that it passes `isIdentifier(expr)` and does the

```
// If the test was a simple identifier, the above check is sufficient
```
check.

Otherwise it's treating it as a nested object, which will fail

tjenkinson commented:
Thanks

> can you add some tests so we can track which cases are currently supported and which are not, and which are potential false errors?

sure I'll try and add some. Didn't want to go to far before figuring out whether this is actually feasible, and what checks we should be doing

tjenkinson commented on src/compiler/checker.ts:
```diff
@@ -44536,31 +44536,39 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 return;
             }
-            const type = location === condExpr ? condType : checkExpression(location);
+            const type = location === inputCondExpr ? condType : checkExpression(location);
```
I think this was a bug before? It's an optimisation to not need to call `checkExpression` to get the `type` and use the provided `condType`, if the location still matched the provided `condExpr`?

tjenkinson commented on tests/baselines/reference/falsinessCallExpressionCoercion.errors.txt:
```diff
@@ -0,0 +1,124 @@
+             ~~~~~~~~~
+!!! error TS2774: This condition will always return true since this function is always defined. Did you mean to call it instead?
+        } else {
+            canAccess()
```
This is erroring, and it seems that `forEachChild` is not finding this node

But if I change the if to `if (!canAccess || true)` then it does find this call

Any idea why it's sometimes ignoring the `else`?

tjenkinson replied:
Actually

```ts
    const item = () => true;

    if (!item || true) {
    } else {
    }
```
doesn't fail when it should so might be a different cause

The issue is though that 

```ts
    const item = () => true;

    if (!item) {
    } else {
      item
    }
```

fails when it shouldn't, whereas

```ts
    const item = () => true;

    if (!item) {
    } else {
    }

    item
```

is fine

tjenkinson replied:
nevermind figured this one out

tjenkinson commented:
hey @gabritto 

I've updated it now to check to see if the variable is referenced, after the point that's being tested, which hopefully rule out a lot of the errors from the open source builds. I initially wanted to check specifically if the fn was called/promise awaited later using `getSignaturesOfType` and `getAwaitedTypeOfPromise`, but in order to do that I needed `type` and to get that called `checkExpression`, but this caused some strange errors. I guess `checkExpression` has side effects

Also added some tests

There is an issue though which I added a failing test for, which is the cause of the `contextualOverloadListFromArrayUnion` errors. Added more info in a [comment on that file](https://github.com/microsoft/TypeScript/pull/60068/files#r1929859488) would be great if you have any ideas

gabritto commented:
@typescript-bot test it

tjenkinson commented:
hey @gabritto what do you think? Looks like performance has taken a bit of a hit? Might need a cache so not checking the same block multiple times?

tjenkinson commented:
Hey @gabritto any update? Wondering if we might need to port this to Go 😅

We just fixed another production bug that would have been caught by this 🤦

tjenkinson commented:
@gabritto no worries 

> We're also still some time away from adding new features to the Go codebase, so it might take a while anyway for this to make it into TS

I’m happy to look at porting this to Go. Is there a Go branch I could raise a pr against? Or would it be better to hold off until after the first Go release is ready?

gabritto commented:
> @gabritto no worries
> 
> > We're also still some time away from adding new features to the Go codebase, so it might take a while anyway for this to make it into TS
> 
> I’m happy to look at porting this to Go. Is there a Go branch I could raise a pr against? Or would it be better to hold off until after the first Go release is ready?

Probably better to wait until the Go version is more stable.
